File: src/services/account_services.py
from src.entities.person import Person
from src.utils.general import AccountInfo, AccountErrors

from src.interfaces.iaccountoperation import IAccountOperation

class AccountOperations:
    def __init__(self, person: Person, operations: IAccountOperation):
        self.person = person
        self.operations = operations

    def update_balance(self, amount: float):
        client_balance = self.person.account.get_balance()
        special_balance = self.person.account.get_special_credit()
        
        if amount > 0: # + 300
            self.operations.credit(amount)
            return AccountInfo.SUCCESS_CREDIT
        
        if client_balance < 0: #special credit
            if (amount * -1) > ((client_balance*-1) + (special_balance)):
                raise AccountErrors.NO_ENOUGH_LIMIT.value
            
            if client_balance + (special_balance*-1) > client_balance + amount:
                raise AccountErrors.NO_ENOUGH_LIMIT.value

            self.operations.debit(amount)
            return AccountInfo.SUCCESS_DEBIT
        
        if (client_balance - (amount*-1)) < 0:
            raise AccountErrors.NO_ENOUGH_MONEY.value        
            
        else:
            self.operations.debit(amount)
            return AccountInfo.SUCCESS_DEBIT

File: go/pkg/services/account_services.go
package services

import (
	"bank-system/pkg/entities"
	"bank-system/pkg/operations"
	"fmt"
	"log"
)

type IAccountService interface {
	UpdateBalance(amount float32) (string, error)
}

type AccountService struct {
	Person    entities.IPerson
	Operation operations.IAccountOperation
}

func (a *AccountService) UpdateBalance(amount float32) (string, error) {
	clientAccount, err := a.Person.GetAccount()
	if err != nil {
		return "", err
	}
	clientBalance := clientAccount.GetBalance()
	clientSpecialBalance := clientAccount.GetSpecialCredit()

	if amount > 0 {
		log.Println("Doing Credit transaction 1")
		a.Operation.Credit(amount)
		return "Success credit", nil
	}

	if clientBalance < 0 {
		if (amount*-1) > ((clientBalance*-1)+clientSpecialBalance) || clientBalance+(clientSpecialBalance*-1) > clientBalance+amount {
			return "No Enough Limit", fmt.Errorf("No Enough Limit")
		}
		a.Operation.Debit(amount)
		log.Println("Doing Debit transaction 1")
		return "Success debit", nil
	}

	if ((clientBalance + clientSpecialBalance) - (amount * -1)) < 0 {
		return "No Enough Limit", fmt.Errorf("No Enough Limit")

	}
	log.Println("Doing Debit transaction 2")
	a.Operation.Debit(amount)
	return "Success debit", nil

}
